tests/integration/Core/Content/Category/SalesChannel/NavigationRouteTest.php

```diff
@@ -239,6 +246,192 @@ public function testInclude(): void
+            }
+        }
+
+        $this->createSeoUrl(
+            'frontend.detail.page',
+            '/detail/' . $productId,
+            'custom-product-url',
+            $productId
+        );
+
+        $response = $this->requestFooterNavigationWithSeoUrls();
+
+        foreach ($response as $category) {
+            if ($category['id'] === $this->ids->get('category3') && $category['linkType'] === CategoryDefinition::LINK_TYPE_PRODUCT) {
+                static::assertStringContainsString('/custom-product-url', $category['seoLink']);
+            }
+        }
+    }
+
+    public function testCategoryInternalLinkHasSeoUrl(): void
+    {
+        $this->getContainer()->get('category.repository')->update([
+            [
+                'id' => $this->ids->get('category3'),
+                'type' => CategoryDefinition::TYPE_LINK,
+                'linkType' => CategoryDefinition::LINK_TYPE_CATEGORY,
+                'internalLink' => $this->ids->get('category2'),
+            ],
+        ], Context::createDefaultContext());
+
+        $response = $this->requestFooterNavigationWithSeoUrls();
+
+        foreach ($response as $category) {
+            if ($category['id'] === $this->ids->get('category3') && $category['linkType'] === CategoryDefinition::LINK_TYPE_CATEGORY) {
+                static::assertNotEmpty($category['seoLink']);
+            }
+        }
+
+        $this->createSeoUrl(
+            'frontend.navigation.page',
+            '/navigation/' . $this->ids->get('category2'),
+            'custom-category-url',
+            $this->ids->get('category2')
+        );
+
+        $response = $this->requestFooterNavigationWithSeoUrls();
+
+        foreach ($response as $category) {
+            if ($category['id'] === $this->ids->get('category3') && $category['linkType'] === CategoryDefinition::LINK_TYPE_CATEGORY) {
+                static::assertStringContainsString('/custom-category-url', $category['seoLink']);
+            }
+        }
+    }
+
+    /**
+     * Helper method to create a pre-configured SEO URL for an entity
+     */
+    private function createSeoUrl(string $routeName, string $pathInfo, string $seoPathInfo, string $entityId): void
+    {
+        $criteria = new Criteria();
+        $criteria->addFilter(new EqualsFilter('foreignKey', $entityId));
+        $criteria->addFilter(new EqualsFilter('routeName', $routeName));
+
+        $existingSeoUrls = $this->getContainer()->get('seo_url.repository')
+            ->search($criteria, Context::createDefaultContext());
+
+        $data = [
+            'salesChannelId' => $this->ids->get('sales-channel'),
+            'routeName' => $routeName,
+            'pathInfo' => $pathInfo,
+            'seoPathInfo' => $seoPathInfo,
+            'isCanonical' => true,
+            'foreignKey' => $entityId,
+        ];
+
+        if ($existingSeoUrls->count() > 0 && $existingSeoUrls->first() !== null) {
+            /** @var SeoUrlEntity $seoUrl */
```

Comment:
Inline annotations are not required in tests, you can easily confirm that it is a SeoUrlEntity via `static::assertInstanceOf(SeoUrlEntity::class, $seoUrl);`